### ISSUE TITLE
Consider caching Spdx-Java-library state

### ISSUE THREAD

**Issue by pmonks** [pmonks/clj-spdx#22]
The Spdx-Java-Library is exceptionally expensive to initialise with the basic reference data (license and exception lists, templates etc.). On my laptop it generally takes more than a minute to pull this information down, parse it, and load it into memory.

There may be a way to cache the library's state to disk (e.g. via serialisation, perhaps using `nippy`) and only incur the cost of initialisation when the lists have changes (which can be determined via an eTag request to the two list URLs).

**Comment by pmonks** [comment]
A quick experiment with `nippy` suggests this may not be easy solely from Clojure:
```
$ clj -Sdeps '{:deps {com.taoensso/nippy {:mvn/version "RELEASE"}}}' -r
user=> (require '[taoensso.nippy :as nippy])
nil
user=> (require '[spdx.licenses :as sl])
nil
user=> (sl/init!)
nil
user=> (require '[spdx.exceptions :as se])
nil
user=> (se/init!)
nil
user=> (require '[spdx.impl.state :as sis])
nil
user=> (nippy/freeze @sis/list-obj)
Execution error (ExceptionInfo) at taoensso.nippy/throw-unfreezable (nippy.clj:999).
Unfreezable type: class org.spdx.library.model.license.ListedLicenses
user=> 
```

**Comment by pmonks** [comment]
Raised https://github.com/spdx/Spdx-Java-Library/issues/193 to see if this can be fixed in the underlying Java library (which is a better place for it anyway).

**Comment by pmonks** [comment]
This is fixed in Spdx-Java-Library v1.1.8.

**Comment by fmjrey** [comment]
Hi @pmonks, sorry to dig up this old closed issue but before creating any new one I'd rather double check with you a couple things.

In the context of generating new clojure projects based on predefined templates, I'm using your library to load a single license from its id in order to generate the license file. Such use case would basically require only a limited amount of licenses at hand. In clojureland that would be EPL-1.0/2.0 and a few other suspects like MIT, Apache, etc.

So here's my question: is there a way to completely bypass the full cache download and only build a local cache as requests come in?

I do think a cache is still useful in our use case, otherwise creating a project from template would *always* require an internet connection. Digging into your past discussions I fear the official java lib always download a full cache in memory at startup, which is why you worked out a persistent local cache to avoid that cost across separate runs.

I also see you are currently working on [urlocal](https://github.com/pmonks/urlocal) so might I guess you're thinking of a cache-as-you-go solution?

For reference see [licenses.clj](https://github.com/fmjrey/project-templates/blob/licenses/src/practicalli/licenses.clj) and accompanying [PR](https://github.com/practicalli/project-templates/pull/37). I've enganged Sean Corfield, author of `deps-new`, about adding this to deps-new. Happy to move our discussion elsewhere if needed.

Edit: I've discovered the SPDX jar actually contains the license data, which should be fit for purpose in our use case. However an incremental cache would be nice too.

**Comment by pmonks** [comment]
Hi @fmjrey!  👋   Sorry for the slow reply - I was on vacation this week.  Oh and absolutely no worries about raising this here (or as a new issue or whatever) - no need to apologise!

> So here's my question: is there a way to completely bypass the full cache download and only build a local cache as requests come in?

tl;dr - don't call any of the `init!` fns in clj-spdx.

Long answer: provided you don't call any of the various `init!` fns, clj-spdx will simply pass through calls to Spdx-Java-Library on demand, which should, in theory, amortise the Java library's initialisation cost over time, and limit it to only retrieving the license / exception data you need for your use case.

**However**

what I've found is that sometimes Spdx-Java-Library seems to fully initialise itself even for a call that's seemingly narrow / limited in scope, so YMMV.  If you happen to notice oddball behaviour like that, please raise issues in the [Spdx-Java-Library repository](https://github.com/spdx/Spdx-Java-Library/issues) - I've worked with the author quite a bit and he's very keen to hear of implementer issues with the library.

Related: I've been meaning to profile Spdx-Java-Library at some point to figure out why it's so slow to initialise.  When I first started developing clj-spdx (before I decided to wrap Spdx-Java-Library instead), I initially pulled down the license and exception lists, matching templates, etc. just using normal Clojure, cached them locally on disk, and parsed them straight into memory, and it was _massively_ faster than Spdx-Java-Library is (like a second or two total, even when the local cache wasn't populated yet).  I'm therefore fairly certain there are still easy performance wins to be found in the Java library, and FWIW I've already contributed a few that sped it up (though clearly not enough!):

* https://github.com/spdx/Spdx-Java-Library/issues/193
* https://github.com/spdx/Spdx-Java-Library/issues/202

A cold start of the Java library takes over a minute on my laptop, which seems outrageous to me, given what it's doing (i.e. basically downloading a couple of thousand fairly small files totaling only 26MB, and parsing around 800 of them).

But I digress...

> I also see you are currently working on [urlocal](https://github.com/pmonks/urlocal) so might I guess you're thinking of a cache-as-you-go solution?

Yeah in fact clj-spdx does _not_ use urlocal, though [I implemented basically the same thing inside Spdx-Java-Library](https://github.com/spdx/Spdx-Java-Library/blob/master/src/main/java/org/spdx/utility/DownloadCache.java) - it made more sense to me to do it at the Java level, so that all JVM language users of Spdx-Java-Library benefit.

> For reference see [licenses.clj](https://github.com/fmjrey/project-templates/blob/licenses/src/practicalli/licenses.clj) and accompanying https://github.com/practicalli/project-templates/pull/37. I've enganged Sean Corfield, author of deps-new, about adding this to deps-new. Happy to move our discussion elsewhere if needed.

Cool!  Love to see folx advocating for adoption of SPDX, especially when it makes it a no-brainer for developers!

And please let me know if you do end up discussing this elsewhere and I'd be happy to join if you think that would be helpful.  Though I should mention that I'm not on the Clojurians Slack (and have no intent to join it), so I would not be able to join any discussions that happen there.  I am active on the [Clojure Discord chat server](https://discord.gg/discljord), if that's a suitable venue for you, though.

> Edit: I've discovered the SPDX jar actually contains the license data, which should be fit for purpose in our use case. However an incremental cache would be nice too.

Yeah the Java library has an "offline" mode, since it's used in regulated scenarios where the JVM doesn't have internet access.  That mode is enabled via the [`org.spdx.useJARLicenseInfoOnly` environment variable](https://github.com/spdx/Spdx-Java-Library?tab=readme-ov-file#configuration-options).  You should be able to easily test your existing code with `org.spdx.useJARLicenseInfoOnly=true` set in the JVM's environment, and see what (if any) difference that makes in terms of initialisation time.

And yes, _in theory_, Spdx-Java-Library's cache is incremental, provided you _don't_ call any of clj-spdx's `init!` fns, though as I say, I've found that sometimes it'll force initialise everything anyway...

**Comment by fmjrey** [comment]
Hi @pmonks, thanks for chiming in.

I only use `id->license` with the option to download the full text. The way you explain it makes me believe there is no determinitic way to know when a full download is triggered. If not calling `init!` is such a gamble then I'd rather not rely on that.

Sounds like your method of downloading the data directly in clojure would better suit our use case. Maybe there could be different ways to initialize `clj-spdx`, e.g. by having a `:license-store` options that specifies `:jar`, `:cache`, `:api`, or `:clj` (urlocal), if that makes sense? 
Edit: that last `:clj` option probably means the java library can't be used though. So maybe that's calling for another lib...

**Comment by pmonks** [comment]
@fmjrey wrote:

> I only use id->license with the option to download the full text. The way you explain it makes me believe there is no determinitic way to know when a full download is triggered. If not calling init! is such a gamble then I'd rather not rely on that.

It's been a while since I ran into that issue, and there've been a few releases of Spdx-Java-Library since then, so I think it's a reasonably safe gamble.

Plus if you do run into unexpected full initialisation I'm sure the author of Spdx-Java-Library would love to know about it - I (and I think he) would consider that a bug.

> Sounds like your method of downloading the data directly in clojure would better suit our use case.

The challenge with that approach is that it wouldn't be possible to initialise Spdx-Java-Library from those files, so it would either incur Spdx-Java-Library's initialisation costs anyway (e.g. the first time `id->info` or whatever is called), or clj-spdx would need parallel Clojure-only implementations of every fn (which defeats the purpose of being a thin wrapper around Spdx-Java-Library).

Furthermore there are some users of clj-spdx who use parts of Spdx-Java-Library that aren't wrapped yet via direct interop (e.g. the SPDX model classes), and their code could start behaving unexpectedly if Spdx-Java-Library isn't properly initialised prior to their interop calls.

But tl;dr - I'd encourage you to try your code without any calls to clj-spdx's (optional) `init!` fns, and if you happen to run into full initialisation, raise an issue (either here or directly in the Spdx-Java-Library repository) and I'll coordinate with Gary to see if we can get to the bottom of what's going on.

As I say I consider this slow initialisation behaviour a pretty strong sign that there's something funky going on in Spdx-Java-Library, but I haven't gotten around to digging into why it's so slow myself yet (and tbh every time I go back to Java-the-language I get PTSD, so I'm probably also guilty of postponing this work longer than I should 😜).

**Comment by pmonks** [comment]
@fmjrey I got curious, so created `test.clj` with this content:
```clojure
(require '[spdx.identifiers :as si])

(si/id->info "Apache-2.0")
(si/id->info "EPL-1.0")
(si/id->info "EPL-2.0")
(si/id->info "MIT")
(si/id->info "GPL-2.0")
```

Then ran:
```shell
$ rm -rf ~/.cache/Spdx-Java-Library
$ time clj -M test.clj
```

The output was:
```text
SLF4J(W): No SLF4J providers were found.
SLF4J(W): Defaulting to no-operation (NOP) logger implementation
SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.

________________________________________________________
Executed in    2.14 secs    fish           external
   usr time    2.42 secs    0.32 millis    2.42 secs
   sys time    0.14 secs    2.04 millis    0.14 secs
```

And then I ran this, just to check how much of Spdx-Java-Library's cache got populated:
```shell
$ ls ~/.cache/Spdx-Java-Library/ | wc -l
      14
```

After a full `init!` (which took over 3 minutes!!) , I see:
```shell
$ ls ~/.cache/Spdx-Java-Library/ | wc -l
    1626
```

So that suggests that just getting the "info" map for a given SPDX identifier is indeed lazy initialising only what's requested, as expected.

**Comment by fmjrey** [comment]
Thanks for checking.  I also validated a similar outcome using my local `deps-new` fork, not calling `init!`, and checking local cache.

BTW, the SLF4J warning now appears when running `deps-new`, which isn't nice. I believe you can completely remove it by replacing `slf4j-api` by `slf4j-nop` in your `deps.edn` (same version info), please check on your side. EDIT:  actually Sean Corfield says it's best to add it in deps-new because clj-spdx is a library...

I see some licenses text have placeholders for things like year and owner. For example the MIT license text contains `Copyright (c) <year> <copyright holders>`.  The SPDX specs describe these as [replaceable text](https://spdx.github.io/spdx-spec/v3.0.1/annexes/license-matching-guidelines-and-templates/#guideline-replaceable-text) and seems to have moved on [from text to XML for templates](https://spdx.github.io/spdx-spec/v3.0.1/annexes/license-matching-guidelines-and-templates/#legacy-text-template-format). Based on the spec wording and skimming through the API I don't believe SPDX at present is intended for generating license files, but instead for comparing and validating conformity with a template. Also I don't believe there is any standardized set of placeholders. Is that your understanding as well?

**Comment by pmonks** [comment]
@fmjrey wrote:

> BTW, the SLF4J warning now appears when running deps-new, which isn't nice. I believe you can completely remove it by replacing slf4j-api by slf4j-nop in your deps.edn (same version info), please check on your side. EDIT: actually Sean Corfield says it's best to add it in deps-new because clj-spdx is a library...

That's correct - clj-spdx is a library, so it's up to downstream users of the library to choose a logging implementation appropriate for their needs.  And those warnings are actually coming from Spdx-Java-Library (which produces logs via [SLF4J](https://slf4j.org/)) - clj-spdx itself does not emit any logging itself.

> I see some licenses text have placeholders for things like year and owner. For example the MIT license text contains Copyright (c) <year> <copyright holders>. The SPDX specs describe these as [replaceable text](https://spdx.github.io/spdx-spec/v3.0.1/annexes/license-matching-guidelines-and-templates/#guideline-replaceable-text) and seems to have moved on [from text to XML for templates](https://spdx.github.io/spdx-spec/v3.0.1/annexes/license-matching-guidelines-and-templates/#legacy-text-template-format).

Correct.  [Spdx-Java-Library still uses the text templates](https://github.com/spdx/spdx-java-core/blob/main/src/main/java/org/spdx/licenseTemplate/SpdxLicenseTemplateHelper.java), I assume for historical reasons (the library predates the XML versions of the matching templates).

If you're interested in obtaining the raw text of a license, or the text rendition of the matching template, you can set the `include-large-text-values?` flag in the call to `id->info`.  For example:
```clojure
(require '[spdx.identifiers :as si])

(si/id->info "MIT" {:include-large-text-values? true})
```
This will emit something like:
```edn
{:id "MIT",
 :name "MIT License",
 :see-also
 ("https://opensource.org/license/mit/"
  "http://opensource.org/licenses/MIT"),
 :fsf-libre? true,
 :osi-approved? true,
 :text
 "MIT License\n\nCopyright (c) <year> <copyright holders>\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of this software and\nassociated documentation files (the \"Software\"), to deal in the Software without restriction, including\nwithout limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the\nfollowing conditions:\n\nThe above copyright notice and this permission notice shall be included in all copies or substantial\nportions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT\nLIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO\nEVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE\nUSE OR OTHER DEALINGS IN THE SOFTWARE.\n",
 :text-template
 "<<beginOptional>>MIT License\n\n<<endOptional>> <<var;name=\"copyright\";original=\"Copyright (c) <year> <copyright holders>  \";match=\".{0,5000}\">>\nPermission is hereby granted, free of charge, to any person obtaining a copy of <<var;name=\"files\";original=\"this software and associated documentation files\";match=\"this\\s+software\\s+and\\s+associated\\s+documentation\\s+files|this\\s+source\\s+file\">> (the \" <<var;name=\"Software1\";original=\"Software\";match=\"Software|Materials\">> \"), to deal in the <<var;name=\"Software2\";original=\"Software\";match=\"Software|Materials\">> without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the <<var;name=\"Software3\";original=\"Software\";match=\"Software|Materials\">> , and to permit persons to whom the <<var;name=\"Software4\";original=\"Software\";match=\"Software|Materials\">> is furnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice<<beginOptional>> (including the next paragraph)<<endOptional>> shall be included in all copies or substantial portions of the <<var;name=\"Software5\";original=\"Software\";match=\"Software|Materials\">> .\n\nTHE <<var;name=\"Software-verb\";original=\"SOFTWARE IS\";match=\"SOFTWARE IS|MATERIALS ARE\">> PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL <<var;name=\"copyrightHolder\";original=\"THE AUTHORS OR COPYRIGHT HOLDERS\";match=\".+\">> BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE <<var;name=\"Software7\";original=\"SOFTWARE\";match=\"SOFTWARE|MATERIALS\">> OR THE USE OR OTHER DEALINGS IN THE <<var;name=\"Software8\";original=\"SOFTWARE\";match=\"SOFTWARE|MATERIALS\">> .\n\n",
 :type :license-id}
```

The `:text` and `:text-template` elements are not included by default since they are large and so (in aggregate) can consume quite a bit of JVM heap.  I've also found them to be less generally useful than the "structured" metadata elements (`:name`, `:see-also`, `:fsf-libre?`, `:osi-approved?`, etc.), so felt that gating them behind an optional flag was a better design approach.

I'm not aware of a way to obtain the XML templates using Spdx-Java-Library (and hence also clj-spdx), presumably because that library doesn't use them (so downloading them would add cost for no benefit), but feel free to raise a new issue here if you'd like to be able to access them, and I can work with Gary to get it added to Spdx-Java-Library (and then wrap it into clj-spdx).

> Based on the spec wording and skimming through the API I don't believe SPDX at present is intended for generating license files, but instead for comparing and validating conformity with a template.

That's correct, though I suspect the `:text` element returned from `id->info` (when the `include-large-text-values?` flag is set) can be used for that purpose anyway, despite SPDX not intending it for that purpose.

> Also I don't believe there is any standardized set of placeholders. Is that your understanding as well?

I'm not 100% sure, but in the templates at least I believe that's correct.  This is largely because license publishers themselves don't follow any particular standard around those placeholders.  [`MIT`](https://opensource.org/license/mit) uses `<year> <copyright holders>` for example, while [`EPL-2.0`](https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.txt) doesn't have placeholders at all - the bottom of the license simply says `You may add additional accurate notices of copyright ownership.` and leaves it up to the author of the project to express their copyright claim correctly.

In general SPDX tries _very_ hard not to introduce _anything_ into license texts, since that can have legal consequences, and one of [the SPDX project's guiding principles](https://spdx.dev/about/overview/) is "SPDX makes no legal interpretations (of licenses or license compliance).".

**Comment by fmjrey** [comment]
Thank you again for all these details. I think for our use case we'll just default to the jar but provide options to let the user decide if they want to use the cache. You can see the code in my [deps-new fork](https://github.com/fmjrey/deps-new).
> clj-spdx itself does not emit any logging itself

I'm surprise that based on your comment in deps.edn you had to add the slf4j-api lib...

**Comment by pmonks** [comment]
> Thank you again for all these details. I think for our use case we'll just default to the jar but provide options to let the user decide if they want to use the cache.

In that case I'd recommend just documenting the clj-spdx / Spdx-Java-Library's default configuration mechanism for doing this i.e. [the `org.spdx.useJARLicenseInfoOnly` JVM property](https://github.com/spdx/Spdx-Java-Library?tab=readme-ov-file#configuration-options) ~~(along with a caveat that the clj-spdx default is the opposite of the Spdx-Java-Library default).~~ [edit] this is incorrect - I was thinking of `org.spdx.downloadCacheEnabled`, where the default _is_ reversed from what Spdx-Java-Library does.

> I'm surprise that based on your comment in deps.edn you had to add the slf4j-api lib...

As the comment on [that line in `deps.edn`](https://github.com/pmonks/clj-spdx/blob/dev/deps.edn#L13) says:

```clojure
; This is required by org.spdx/java-spdx-library, yet somehow isn't coming through transitively
```

Spdx-Java-Library _does_ emit logging, so even though clj-spdx doesn't, the SLF4J library is required (and wasn't coming through transitively for some reason, though it's been a long time since I retested that).

**Comment by fmjrey** [comment]
I'm not so sure it's a good idea to expose other SPDX env vars in `deps-new`. I learned about them when seeing your static initilizer in [state.clj](https://github.com/pmonks/clj-spdx/blob/dev/src/spdx/impl/state.clj). You certainly had good reasons for that overriding behavior, though overriding the env var leaves no choice to the user. Exposing other SPDX env var while one is always overriden by `clj-spdx` is kind of confusing from a user point of view. Feels like a leaky abstraction to expose some settings of an inner library.

Also env vars make settings out of band and system wide, making them local require some wrapping logic like a shell script or cli option, or in your case a static initializer. The idiomatic way in clojure is to have a main option or an entry in an option map, the latter being what I opted for. Makes the config more visible in the code and facilitates testing. They can make sense for a long running app or service, but for a short running tool like deps-new I'm not so sure.

**Comment by pmonks** [comment]
@fmjrey wrote:

> I'm not so sure it's a good idea to expose other SPDX env vars in `deps-new`.

For better or worse that's how Spdx-Java-Library is designed, and I chose not to invent a different scheme on top of that in clj-spdx, though of course you could do so in `deps-new`.

> I learned about them when seeing your static initilizer in [state.clj](https://github.com/pmonks/clj-spdx/blob/dev/src/spdx/impl/state.clj). You certainly had good reasons for that overriding behavior, though overriding the env var leaves no choice to the user.

It doesn't override any user env vars though - it only sets one of them (to the opposite default of Spdx-Java-Library) _if it isn't already set_ (see [the conditional on line 17](https://github.com/pmonks/clj-spdx/blob/dev/src/spdx/impl/state.clj#L17)).

Also, as is obvious from that line, I was misremembering in my previous reply where I said that clj-spdx reverses the default value of the `org.spdx.useJARLicenseInfoOnly` JVM property.  It's actually the `org.spdx.downloadCacheEnabled` JVM property where clj-spdx has the opposite default (enabled) to Spdx-Java-Library's default (disabled).  Of course that's only changing the _default_ - the user's own env vars _always_ take precedence.

> Exposing other SPDX env var while one is always overriden by `clj-spdx` is kind of confusing from a user point of view. Feels like a leaky abstraction to expose some settings of an inner library.

That would indeed be confusing if that's how clj-spdx worked, but it isn't - the user's own env vars _always_ take precedence (over both clj-spdx's defaults and Spdx-Java-Library's defaults).

> Also env vars make settings out of band and system wide, making them local require some wrapping logic like a shell script or cli option, or in your case a static initializer. The idiomatic way in clojure is to have a main option or an entry in an option map, the latter being what I opted for. Makes the config more visible in the code and facilitates testing. They can make sense for a long running app or service, but for a short running tool like deps-new I'm not so sure.

You're welcome to [raise an issue against Spdx-Java-Library](https://github.com/spdx/Spdx-Java-Library/issues/new) if you'd like to propose a different configuration mechanism.  As mentioned above, that's simply how Spdx-Java-Library was designed, and I (deliberately) chose not to invent a different configuration scheme in clj-spdx, since it's explicitly intended to be as thin a wrapper around the underlying Java library as possible.

**Comment by fmjrey** [comment]
> It doesn't override any user env vars though - it only sets one of them (to the opposite default of Spdx-Java-Library) if it isn't already set (see [the conditional on line 17](https://github.com/pmonks/clj-spdx/blob/dev/src/spdx/impl/state.clj#L17)).
(...)
> That would indeed be confusing if that's how clj-spdx worked, but it isn't - the user's own env vars always take precedence (over both clj-spdx's defaults and Spdx-Java-Library's defaults).

Maybe it's a bit of a stretch, my thinking was that if the user intentionally does not set `org.spdx.downloadCacheEnabled`  then `clj-spdx` is in effect overriding the user's intention. Though considering how you described the performance issues I don't see how one would not want it set. You even wrote some java to solve that issue!

> As mentioned above, that's simply how Spdx-Java-Library was designed, and I (deliberately) chose not to invent a different configuration scheme in clj-spdx, since it's explicitly intended to be as thin a wrapper around the underlying Java library as possible.

Makes sense and I can't say I'd have done better. I'm really glad you worked out the cache like you did. For the SPDX library it also makes sense to have these env vars, though I'd argue for providing programmatic ways to change these settings with some optional parameters, which indeed would override the env vars (saying that though I did not check the java lib for such options). In the case of `deps-new` I think the context is different, it's a cli tool that is run from time to time and should not take much time carrying out its work.

**Comment by fmjrey** [comment]
Just thought of a scenario that options on top of env var would make possible/easier: first check with the jar option, if not found check SPDX cache/web.

**Comment by pmonks** [comment]
> Maybe it's a bit of a stretch, my thinking was that if the user intentionally does not set org.spdx.downloadCacheEnabled then clj-spdx is in effect overriding the user's intention. Though considering how you described the performance issues I don't see how one would not want it set. You even wrote some java to solve that issue!

Yep.  I advocated for the download cache to be enabled by default (as it is in clj-spdx), but at the time the author of Spdx-Java-Library wasn't comfortable with that as the default.  If you were to raise that as an issue against Spdx-Java-Library he might reconsider that decision?

> For the SPDX library it also makes sense to have these env vars, though I'd argue for providing programmatic ways to change these settings with some optional parameters, which indeed would override the env vars (saying that though I did not check the java lib for such options). In the case of deps-new I think the context is different, it's a cli tool that is run from time to time and should not take much time carrying out its work.

Keep in mind they're not env vars - they're [JVM system properties](https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html), which _can_ be set programmatically (in fact that's exactly what clj-spdx does on [line 18](https://github.com/pmonks/clj-spdx/blob/dev/src/spdx/impl/state.clj#L18) to enable the download cache by default).

If deps-new was to set those JVM systems properties as you'd like first, _before_ requiring clj-spdx, you'd get the behaviour you like without any work required of the end user.  If you did that, I'd suggest following the same approach I took in clj-spdx - check if the properties are set first, and only force them to your desired values if they're not set.  That way users can always override deps-new's defaults (as well as clj-spdx's and Spdx-Java-Library's).

> Just thought of a scenario that options on top of env var would make possible/easier: first check with the jar option, if not found check SPDX cache/web.

The JAR always includes the (potentially out of date) assets ([they're vendored into the source repository as classpath resources](https://github.com/spdx/Spdx-Java-Library/tree/master/resources/stdlicenses)), so the fallback to the web wouldn't ever occur in this approach.

Spdx-Java-Library mostly provides that option for use cases where the code doesn't have access to the internet - the library is used in some regulated environments where the machines it's running on are air-gapped.  In fact while confirming my recollection of how Spdx-Java-Library does this, I noticed that [some files are _not_ currently being vendored](https://github.com/spdx/Spdx-Java-Library/issues/396), which will cause some Spdx-Java-Library methods (and hence also some clj-spdx functions) to fail unexpectedly when `org.spdx.useJARLicenseInfoOnly` is `true`.

**Comment by fmjrey** [comment]
> Keep in mind they're not env vars - they're JVM system properties, which can be set programmatically

oh, thanks for the reminder.  The whole time I've discussed with you I really had env vars in mind. I guess it's the [java lib README](https://github.com/spdx/Spdx-Java-Library?tab=readme-ov-file#configuration-options) that set me on this path when it says these must be set once before the first call as changing them afterward has no effect.  So yes they can be changed during execution (but the java lib only read them when initializing), they're not system wide but JVM wide, and it's still awkward to change before a call, so the idea of using options when placing a call still makes sense.  Actually rereading [Spdx-Java-Library#193](https://github.com/spdx/Spdx-Java-Library/issues/193) I see you argued for that.

> If deps-new was to set those JVM systems properties as you'd like first, before requiring clj-spdx, you'd get the behaviour you like without any work required of the end user.

before requiring sounds awkward to do. I suppose it's the way the java lib works that made you have a static initializer.

> Yep. I advocated for the download cache to be enabled by default (as it is in clj-spdx), but at the time the author of Spdx-Java-Library wasn't comfortable with that as the default. If you were to raise that as an issue against Spdx-Java-Library he might reconsider that decision?

Would it not be best to advocate for call options and mention how it would solve the default issue?

**Comment by fmjrey** [comment]
BTW Sean Corfield would have liked to comment here but he says he's blocked?

**Comment by pmonks** [comment]
> before requiring sounds awkward to do. I suppose it's the way the java lib works that made you have a static initializer.

That's correct. I had to set the JVM system properties before the call that initialises Spdx-Java-Library, because once that call is made and Spdx-Java-Library has configured itself accordingly, it doesn't look at the system properties again (and there's no way to "reset" it and force it to re-initialise itself, short of restarting the JVM).

> Would it not be best to advocate for call options and mention how it would solve the default issue?

Perhaps, though call options aren't idiomatic in Java, and Spdx-Java-Library is primarily intended for Java-the-language users who are familiar with the kinds of OO design approaches it uses.  clj-spdx just happens to (mostly) hide all that OO Java stuff behind a more idiomatic Clojure façade, but there are some places (like initialisation) where it's difficult to completely hide it without compromising configurability (which I wasn't willing to do).

Of course you're welcome to suggest whatever improvements you like to the Spdx-Java-Library author, but it is, at heart, a fairly typical Java / OO library, with typical Java / OO design patterns in use.

**Comment by fmjrey** [comment]
> Of course you're welcome to suggest whatever improvements you like to the Spdx-Java-Library author, but it is, at heart, a fairly typical Java / OO library, with typical Java / OO design patterns in use.

Yeah, I've been spoiled with clojure for a while and easily forget how java is... cumbersome?  I guess if you want some per-call config compatible with spring DI and allow for clojure integration it probably means passing some config object, which implies some important changes throughout the java code, including allowing for more than one license store at any time.  That being said all current JVM properties relate to storage so changes would be well contained.

For now my code has a `:licences` option:

- `:jar`, to get license info from the SPDX jar (default behavior), which sets `org.spdx.useJARLicenseInfoOnly` if not already set

- `:cache`, to get license info from SPDX API and build an incremental cache (does nothing)

- `:full-cache`, to get license info from a full local cache of all SPDX licenses, which adds a noticeable delay upon first run because it calls `init!`.

**Comment by pmonks** [comment]
Yeah nice - that sounds like a great approach!

Also, I dug in a little more to the Spdx-Java-Library performance problem yesterday, and discovered it had a cache stampede problem, which now has a fix: https://github.com/spdx/Spdx-Java-Library/pull/397

It's hard for me to reliably test as my internet seems to be on the fritz, but the best improvement I've seen with cl-spdx's `init!` functions was a reduction in elapsed time from ~3 minutes to ~20 seconds.
